libraries/AP_Relay/AP_Relay.cpp

```diff
@@ -229,6 +281,55 @@ bool AP_Relay::send_relay_status(const GCS_MAVLINK &link) const
 }
 #endif  // AP_MAVLINK_MSG_RELAY_STATUS_ENABLED
 
+void AP_Relay::update()
+{
+    for (uint8_t instance = 0; instance < AP_RELAY_NUM_RELAYS; instance++) {
+        const int8_t pin = _params[instance].pin;
+        if (pin == -1) {
+            // no valid pin to set it on, skip it
+            continue;
+        }
+
+        const AP_Relay_Params::Function function = _params[instance].function;
+        if (function <= AP_Relay_Params::Function::none || function >= AP_Relay_Params::Function::num_functions) {
+            // invalid function, skip it
+            continue;
+        }
+
+        bool value;
+        if (function == AP_Relay_Params::Function::relay) {
+            if (_relay_by_instance_has_been_asserted.get(instance)) {
+                value = _relay_by_instance.get (instance);
+            } else {
+                continue; // still in the default no change state
+            }
+        } else {
+            value = _desired_state.get((uint16_t)function);
+        }
+        const uint8_t old_pin_states = _pin_states;
+        _pin_states = value ? _pin_states | (1U<<instance) : _pin_states & ~(1U<<instance);
+        if (old_pin_states != _pin_states) {
+#if AP_SIM_ENABLED
+            if (!(AP::sitl()->on_hardware_relay_enable_mask & (1U << instance))) {
+                continue;
+            }
+#endif
+            hal.gpio->pinMode(pin, HAL_GPIO_OUTPUT);
+            hal.gpio->write(pin, value);
+        }
+    }
+
+    // log if we changed anything
+    const uint32_t now = AP_HAL::millis();
+    if ((_pin_states != _last_logged_pin_states) && (now - _last_log_ms > 10)) {
+       AP::logger().Write("RELY", "TimeUS,State", "s-", "F-", "QB",
```

Comment:
I think we should log when we do the set/toggle call

Reply:
the RELY log should be an instance specific log message, so it is a log per pin, not all the pin states

Reply:
Done.